tasks/build/download_node_builds.js

```diff
@@ -1,84 +1,134 @@
+    };
+
+    return checkHashFromFile(path, expected).then(([passed, actual]) => {
+      if (passed) grunt.log.ok(`${platform.name} shasum verified`);
+      else grunt.log.error(`${platform.name} shasum check failed`);
+      return passed;
+    });
+  };
 
-      if (platform.win) {
-        await fromNode(cb => {
-          resp
-          .pipe(createWriteStream(resolve(downloadDir, 'node.exe')))
-          .on('error', cb)
-          .on('finish', cb);
-        });
-      } else {
-        await fromNode(cb => {
-          resp
-          .pipe(createGunzip())
-          .on('error', cb)
-          .pipe(new Extract({ path: downloadDir, strip: 1 }))
-          .on('error', cb)
-          .on('end', cb);
-        });
+  const getNodeBuild = (platform) => {
+    const finalDir = platform.nodeDir;
+    const downloadDir = `${finalDir}.temp`;
+    const file = basename(platform.nodeUrl);
+
+    return wreckGet(platform.nodeUrl)
+    .then(([resp, payload]) => {
+      if (resp.statusCode !== 200) {
+        throw new Error(`${platform.nodeUrl} failed with a ${resp.statusCode} response`);
       }
+      const bytes = parseInt(resp.headers['content-length'], 10) || 'unknown number of';
+      const mb = ((bytes / 1024) / 1024).toFixed(2);
+      grunt.log.ok(`downloading ${platform.name} - ${mb} mb`);
+      grunt.file.mkdir(downloadDir);
+      return payload;
+    })
+    .then(payload => {
+      return writeFileAsync(resolve(downloadDir, file), payload);
+    })
+    .then(() => {
+      return renameAsync(downloadDir, finalDir);
+    });
 
-      await fromNode(cb => {
-        rename(downloadDir, finalDir, cb);
+  };
+
+  const extractNodeBuild = async (platform) => {
+    const file = basename(platform.nodeUrl);
+    const path = resolve(platform.nodeDir, file);
+
+    await fromNode(cb => {
+      createReadStream(path)
+      .pipe(createGunzip())
+      .on('error', (err) => {
+        cb(new Error(`error extracting: ${err.message}`));
+      })
+      .pipe(new Extract({path: platform.nodeDir, strip: 1}))
+      .on('error', (err) => {
+        cb(new Error(`error extracting: ${err.message}`));
+      })
+      .on('end', () => {
+        cb(null, true);
       });
-    }());
+    });
+  };
 
-    activeDownloads.push(platform);
+  let downloadFailed = false;
+  const shouldDownload = async (platform, downloadCounter) => {
+    if (downloadCounter === downloadLimit) {
+      downloadFailed = true;
+      return false;
+    }
+    if (!grunt.file.isDir(platform.nodeDir)) {
+      return true;
+    }
+    return !(await checkShaSum(platform));
+  };
 
-    var bytes = parseInt(resp.headers['content-length'], 10) || 'unknown number of';
-    var mb = ((bytes / 1024) / 1024).toFixed(2);
-    grunt.log.ok(`downloading ${platform.name} - ${mb} mb`);
+  const start = async (platform) => {
+    let downloadCounter = 0;
+    while (await shouldDownload(platform, downloadCounter++)) {
```

Comment:
It seems strange that `start` is responsible for maintaining a counter that's used in `shouldDownload`. If you simplify `shouldDownload` as I recommended above, we might be able to get rid of it entirely and simply write this as:

`while (!(await checkShaSum(platform)) && (++downloadCounter < downloadLimit)) {`